File: src/worker/molecule_rdkit.cpp

#include <iostream>

#include <GraphMol/GraphMol.h>

#include <GraphMol/SmilesParse/SmilesParse.h>
#include <GraphMol/FileParsers/FileParsers.h>
#include <GraphMol/FileParsers/MolSupplier.h>

#include <GraphMol/MolOps.h>

#include <GraphMol/DistGeomHelpers/Embedder.h>
#include <GraphMol/ForceFieldHelpers/MMFF/MMFF.h>

#include <GraphMol/MolTransforms/MolTransforms.h>

#include <ForceField/MMFF/Params.h>
#include <ForceField/MMFF/DistanceConstraint.h>
#include <ForceField/MMFF/AngleConstraint.h>
#include <ForceField/MMFF/TorsionConstraint.h>
#include <ForceField/MMFF/PositionConstraint.h>

#include <ForceField/ForceField.h>

#include <sstream>


std::string floattostring(double value)
{
	std::ostringstream strs;
	strs << value;
	std::string str = strs.str();
	return str;
}


void print_molecule(RDKit::ROMOL_SPTR mol, std::string name = "")
{
	// Print molecule in XYZ format

	unsigned int n_atoms = mol->getNumAtoms();
	RDKit::Conformer &conf = mol->getConformer();

	std::cout << n_atoms << "\n" << name << "\n";

	for (unsigned int i=0; i<n_atoms; i++)
	{
		RDKit::Atom *atom = mol->getAtomWithIdx(i);
		RDGeom::Point3D pos = conf.getAtomPos(i);
		std::cout << atom->getAtomicNum() << " ";
		std::cout << pos.x << " ";
		std::cout << pos.y << " ";
		std::cout << pos.z << " ";
		std::cout << "\n";
	}

}


double example_worker(int nexample)
{

	std::string sdf_filename = "sdf/pentane.sdf";

	// filename, sanitize, remove h
	RDKit::SDMolSupplier mol_supplier( sdf_filename, true, false);
	RDKit::ROMOL_SPTR mol( mol_supplier.next() );

	unsigned int n_atoms = mol->getNumAtoms();
	RDKit::Conformer &conf = mol->getConformer();

	double angle;
	double add_angle = 30.0;

	// Begin
	// angle = MolTransforms::getDihedralDeg(conf, 0, 1, 5, 8);
	// print_molecule(mol, floattostring(angle));

	// Setup forcefield
	RDKit::MMFF::MMFFMolProperties *mmffMolProperties = new RDKit::MMFF::MMFFMolProperties(*mol);
    ForceFields::ForceField *ff = RDKit::MMFF::constructForceField(*mol, mmffMolProperties);
    ff->initialize();

	// std::cout << ff->contribs().size() << "\n";

	// angle = MolTransforms::getDihedralDeg(conf, 0, 1, 5, 8);
	// print_molecule(mol, floattostring(angle));

	for (unsigned int n=0; n<nexample; n++)
	{

		angle = angle + add_angle;

		// Get current angle
		// double angle1 = MolTransforms::getDihedralDeg(conf, 0, 1, 5, 8);

		// Set new angle
		MolTransforms::setDihedralDeg(conf, 0, 1, 5, 8, angle);

		// Set constraint
		ForceFields::MMFF::TorsionConstraintContrib *tc;
		tc = new ForceFields::MMFF::TorsionConstraintContrib(ff, 0, 1, 5, 8, angle, angle, 1.0e10);
		ff->contribs().push_back(ForceFields::ContribPtr(tc));

		// double angle2 = MolTransforms::getDihedralDeg(conf, 0, 1, 5, 8);

		// Minimze
		// minimize(unsigned int maxIts = 200, double forceTol = 1e-4, double energyTol = 1e-6);
		ff->minimize(200, 1e-2, 1e-2);

		// double angle3 = MolTransforms::getDihedralDeg(conf, 0, 1, 5, 8);

		// Remove last constraint
		ff->contribs().pop_back();

		// Minimize
		// ff->minimize(200, 1e-2, 1e-2);

		// double angle4 = MolTransforms::getDihedralDeg(conf, 0, 1, 5, 8);

		// std::cout << "N " << n << " ";
		// std::cout << angle1 << " ";
		// std::cout << angle2 << " ";
		// std::cout << angle3 << " ";
		// std::cout << angle4 << " ";
		// std::cout << "\n";

	}

	// std::cout << ff->contribs().size() << "\n";

	// angle = MolTransforms::getDihedralDeg(conf, 0, 1, 5, 8);
	// print_molecule(mol, floattostring(angle));

	delete mmffMolProperties;
	// delete tc; // Delete via pop_back() ?

}


int main( int argc , char **argv ) {

	example_worker(1000);

	return 0;
}

File: src/worker/molecule_rdkit.py

import rmsd

import numpy as np
import time

from rdkit import Chem
from rdkit.Chem import AllChem
from rdkit.Chem import ChemicalForceFields

import matplotlib.pyplot as plt

ATOM_LIST = [x.strip() for x in [
    'h ', 'he', \
    'li', 'be', 'b ', 'c ', 'n ', 'o ', 'f ', 'ne', \
    'na', 'mg', 'al', 'si', 'p ', 's ', 'cl', 'ar', \
    'k ', 'ca', 'sc', 'ti', 'v ', 'cr', 'mn', 'fe', 'co', 'ni', 'cu', \
    'zn', 'ga', 'ge', 'as', 'se', 'br', 'kr',  \
    'rb', 'sr', 'y ', 'zr', 'nb', 'mo', 'tc', 'ru', 'rh', 'pd', 'ag', \
    'cd', 'in', 'sn', 'sb', 'te', 'i ', 'xe',  \
    'cs', 'ba', 'la', 'ce', 'pr', 'nd', 'pm', 'sm', 'eu', 'gd', 'tb', 'dy', \
    'ho', 'er', 'tm', 'yb', 'lu', 'hf', 'ta', 'w ', 're', 'os', 'ir', 'pt', \
    'au', 'hg', 'tl', 'pb', 'bi', 'po', 'at', 'rn', \
    'fr', 'ra', 'ac', 'th', 'pa', 'u ', 'np', 'pu']]


def get_atom(atom):
    """

    *

    """
    atom = atom.lower()
    return ATOM_LIST.index(atom) + 1


def clockwork(n, debug=False):
    """

    get start, step size and no. of steps from clockwork resolution n

    """

    if n == 1:
        start = 0
        step = 0
        n_steps = 0

    else:
        start = 360.0 / 2 ** (n-1)
        step = 360.0 / (2**(n-2))
        n_steps = 2**(n-2)

    if debug:
        print(n, step, n_steps, start)

    return start, step, n_steps


def get_angles(n, debug=False):
    """

    return angles to check at clockwork resolution n

    """

    start, step, n_steps = clockwork(n)
    if n_steps > 1:
        angles = np.arange(start, start+step*n_steps, step)
    else:
        angles = [start]

    return angles


def compare_positions(pos, pos_list, threshold=0.005):
    """

    views_list
    threshold_list

    iterate over views
    - hydrogen views on carbons (Based on bonds)
    - heavy atoms

    """

    for posc in pos_list:
        comparision = rmsd.kabsch_rmsd(pos, posc)

        if comparision < threshold:
            return False

    return True


def align(q_coord, p_coord):
    """

    align q and p.

    return q coord rotated

    """

    U = rmsd.kabsch(q_coord, p_coord)
    q_coord = np.dot(q_coord, U)

    return q_coord


def scan_torsion(resolution, unique_conformers=[]):
    """

    """

    # Load the molecule
    sdf_filename = "sdf/pentane.sdf"
    suppl = Chem.SDMolSupplier(sdf_filename, removeHs=False, sanitize=True)
    mol = next(suppl)

    # Get molecule information
    # n_atoms = mol.GetNumAtoms()
    atoms = mol.GetAtoms()
    # atoms = [atom.GetAtomicNum() for atom in atoms]
    atoms = [atom.GetSymbol() for atom in atoms]

    # Origin
    conformer = mol.GetConformer()
    origin = conformer.GetPositions()
    origin -= rmsd.centroid(origin)

    # Dihedral angle
    a = 0
    b = 1
    c = 5
    d = 8

    # Origin angle
    origin_angle = Chem.rdMolTransforms.GetDihedralDeg(conformer, a, b, c, d)

    # Setup forcefield
    mp = ChemicalForceFields.MMFFGetMoleculeProperties(mol)
    ff = ChemicalForceFields.MMFFGetMoleculeForceField(mol, mp)

    # Define all angles to scan
    angles = get_angles(resolution)

    debug_file = "test.xyz"
    debug_file = open(debug_file, 'a+')

    if len(unique_conformers) == 0:
        xyz = rmsd.calculate_rmsd.set_coordinates(atoms, origin)
        debug_file.write(xyz)
        debug_file.write("\n")

        unique_conformers = [origin]

    for angle in angles:

        # Reset position
        for i, pos in enumerate(origin):
            conformer.SetAtomPosition(i, pos)

        # Set clockwork angle
        Chem.rdMolTransforms.SetDihedralDeg(conformer, a, b, c, d, origin_angle + angle)

        # Setup constrained ff
        ffc = ChemicalForceFields.MMFFGetMoleculeForceField(mol, mp)
        ffc.MMFFAddTorsionConstraint(a, b, c, d, False,
                                     origin_angle+angle, origin_angle + angle, 1.0e10)
        ffc.Minimize(maxIts=1000, energyTol=1e-2, forceTol=1e-3)

        # angle1 = Chem.rdMolTransforms.GetDihedralDeg(conformer, a, b, c, d)

        ff.Minimize(maxIts=1000, energyTol=1e-2, forceTol=1e-4)

        # angle2 = Chem.rdMolTransforms.GetDihedralDeg(conformer, a, b, c, d)

        pos = conformer.GetPositions()
        pos -= rmsd.centroid(pos)

        print("debug", len(unique_conformers))

        unique = compare_positions(pos, unique_conformers)

        if not unique:
            continue

        pos = align(pos, origin)

        unique_conformers.append(pos)

        xyz = rmsd.calculate_rmsd.set_coordinates(atoms, pos)
        debug_file.write(xyz)
        debug_file.write("\n")

        print(angle, unique)

    debug_file.close()

    return unique_conformers


def example_worker(N):

    sdf_filename = "sdf/pentane.sdf"

    suppl = Chem.SDMolSupplier('sdf/pentane.sdf', removeHs=False, sanitize=True)
    mol = next(suppl)

    n_atoms = mol.GetNumAtoms()
    conf = mol.GetConformer()

    add_angle = 30.0

    mp = ChemicalForceFields.MMFFGetMoleculeProperties(mol)

    a = 0
    b = 1
    c = 5
    d = 8

    # Start angle
    angle = Chem.rdMolTransforms.GetDihedralDeg(conf, a, b, c, d)

    ff = ChemicalForceFields.MMFFGetMoleculeForceField(mol, mp)

    pos = ff.Positions()
    pos = np.array(pos)
    pos = pos.reshape((n_atoms, 3))

    # print(pos)

    # help(ff)
    # quit()

    for n in range(N):

        angle = angle + add_angle

        ffc = ChemicalForceFields.MMFFGetMoleculeForceField(mol, mp)
        # ffc.MMFFAddTorsionConstraint(a, b, c, d, False, angle+add_angle, angle+add_angle, 1.0e10)
        Chem.rdMolTransforms.SetDihedralDeg(conf, a, b, c, d, angle + add_angle)
        ffc.Minimize(maxIts=200, energyTol=1e-2, forceTol=1e-2)

        # ff = ChemicalForceFields.MMFFGetMoleculeForceField(mol, mp)
        # ff.Minimize(maxIts=200, energyTol=1e-2, forceTol=1e-2)


    return

if __name__ == '__main__':

    # (clockwork(1))
    # (clockwork(2))
    # (clockwork(3))
    # (clockwork(4))
    # clockwork(5, debug=True)
    # (clockwork(6))

    with open("test.xyz", 'w') as f:
        pass

    unique_conformers = scan_torsion(2)
    unique_conformers = scan_torsion(3, unique_conformers=unique_conformers)
    unique_conformers = scan_torsion(4, unique_conformers=unique_conformers)
    unique_conformers = scan_torsion(5, unique_conformers=unique_conformers)
    unique_conformers = scan_torsion(6, unique_conformers=unique_conformers)
    unique_conformers = scan_torsion(7, unique_conformers=unique_conformers)
    unique_conformers = scan_torsion(8, unique_conformers=unique_conformers)
    unique_conformers = scan_torsion(9, unique_conformers=unique_conformers)
    unique_conformers = scan_torsion(10, unique_conformers=unique_conformers)
    unique_conformers = scan_torsion(11, unique_conformers=unique_conformers)


    quit()
    (clockwork(1))
    (clockwork(2))
    (clockwork(3))
    (clockwork(4))
    (clockwork(5))
    (clockwork(6))

    quit()
    n_list = range(10, 1000, 30)

    xaxis = []
    yaxis = []

    for N in n_list:

        start = time.time()

        example_worker(N)

        end = time.time()

        print(N, end-start)

        ratio = float(N) / (end-start)

        xaxis.append(N)
        yaxis.append(ratio)


    plt.plot(xaxis, yaxis, "k.-")
    plt.savefig("scale_worker_python")

